### PR TITLE
Report Join Server's default JoinEUI

### LINKED ISSUE #4840: Report Join Server's default JoinEUI
#### Summary

Report default JoinEUI

#### Why do we need this?

To make sure that JoinEUIs make sense and for future compatibility with interoperability between network servers.

#### What is already there? What do you see now?

Join Servers are configured with JoinEUI prefixes.

#### What is missing? What do you want to see?

A default JoinEUI within the configured prefix that can be retrieved via a new rpc.

#### How do you propose to implement this?

- Add config to JS with default JoinEUI
- Add `Js` service with rpc `GetDefaultJoinEUI()`
- Implement in JS
- Add CLI support

#### How do you propose to test this?

Simple unit test

#### Can you do this yourself and submit a Pull Request?

Yes, can also review

### PR DESCRIPTION
**Pull request by KrishnaIyer**
#### Summary

Closes #4840 

#### Changes

- Add `DefaultJoinEUI` config and default value.
- Add RPC `GetDefaultJoinEUI` and implementation.
- Add CLI
- Tests
- Fix small CLI panic along the way.
- Docs; https://github.com/TheThingsIndustries/lorawan-stack-docs/pull/693

#### Testing

UT + Local

1. Without a default JoinEUI set in the stack config
```
$ ./ttn-lw-cli end-devices create --application-id test-app < test.json
WARN	Using insecure connection to OAuth server
WARN	Using insecure connection to API
INFO	JoinEUI empty but defaults flag is set, fetch default JoinEUI of the Join Server	{"join_server_address": "localhost:1884"}
INFO	successfully obtained Join Server's default Join EUI	{"default_join_eui": "70B3D57ED0000000"}
{
  "ids": {
    "device_id": "my-device-1",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "0102030405060709",
    "join_eui": "70B3D57ED0000000"
  },
  "created_at": "2021-12-22T10:29:39.394Z",
  "updated_at": "2021-12-22T10:29:39.477996Z",
  "name": "My Device",
  "description": "Living room temperature sensor",
  "attributes": {},
  "network_server_address": "localhost",
  "application_server_address": "localhost",
  "join_server_address": "localhost",
  "lorawan_version": "MAC_V1_0_2",
  "lorawan_phy_version": "PHY_V1_0_2_REV_B",
  "frequency_plan_id": "EU_863_870_TTN",
  "supports_join": true,
  "root_keys": {
    "app_key": {
      "key": "01020304050607080102030405060708"
    }
  }
}
```
2. With a default JoinEUI set in the stack config
```
$ ./ttn-lw-cli end-devices create --application-id test-app < test.json
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
INFO  JoinEUI empty but defaults flag is set, fetch default JoinEUI of the Join Server  {"join_server_address": "localhost:1884"}
INFO  successfully obtained Join Server's default Join EUI  {"default_join_eui": "1122334455667788"}
{
  "ids": {
    "device_id": "my-device-2",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "010203040506070A",
    "join_eui": "1122334455667788"
  },
  "created_at": "2021-12-22T10:31:41.144Z",
  "updated_at": "2021-12-22T10:31:41.643197Z",
  "name": "My Device",
  "description": "Living room temperature sensor",
  "attributes": {},
  "network_server_address": "localhost",
  "application_server_address": "localhost",
  "join_server_address": "localhost",
  "lorawan_version": "MAC_V1_0_2",
  "lorawan_phy_version": "PHY_V1_0_2_REV_B",
  "frequency_plan_id": "EU_863_870_TTN",
  "supports_join": true,
  "root_keys": {
    "app_key": {
      "key": "01020304050607080102030405060708"
    }
  }
}
```

##### Regressions

NA

#### Notes for Reviewers

1. I've set `70b3d57ed0000000` as the [`default` value](https://github.com/TheThingsNetwork/lorawan-stack/blob/065dcf2db8db3ab620dec14b7adeb978075f59c0/cmd/internal/shared/joinserver/config.go#L27) of the default JS Join EUI, so this gets shipped with the stack. We advertise this freely in our docs and I think this is better than leaving it empty. If you have another default value here or if we need to force the setting of this, let me know.
~2. The[ JoinEUI message](https://github.com/TheThingsNetwork/lorawan-stack/blob/065dcf2db8db3ab620dec14b7adeb978075f59c0/api/joinserver.proto#L301-L303) is now defined for the entire API. I don't know if we may need a JoinEUI message somewhere in the future for another purpose? Should I make this more specific like `JSJoinEUI` or something to prevent collisions?~

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/joinserver.proto [hunk 1]
```diff
@@ -298,11 +298,22 @@ message JoinEUIPrefixes {
   repeated JoinEUIPrefix prefixes = 1;
 }
 
+message JoinEUI {
```
Comment (adriansmares) [inline api/joinserver.proto]: `GetDefaultJoinEUIResponse` is more future proof I believe, and less collision prone.

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 2]
```diff
@@ -516,7 +531,7 @@ var (
 					return err
 				}
 				logger.WithField("dev_eui", devEUIResponse.DevEui.String()).
-					Debug("successfully obtained DevEUI")
+					Info("successfully obtained DevEUI")
```
Comment (adriansmares) [inline cmd/ttn-lw-cli/commands/end_devices.go]: Nit: can you fix the casing for both logs ?

**Review by adriansmares — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -11,6 +11,8 @@ For details about compatibility between different releases, see the **Commitment
 
 ### Added
 
+- Support reading the Join Server's default JoinEUI and using this in the CLI for end device creation.
```
Comment (htdvisser) [inline CHANGELOG.md:14]: - Please mention the actual API that was added
- Please mention the configuration that was added

File: pkg/joinserver/config.go [hunk 4]
```diff
@@ -22,5 +22,6 @@ type Config struct {
 	Keys                          KeyRegistry                          `name:"-"`
 	ApplicationActivationSettings ApplicationActivationSettingRegistry `name:"-"`
 	JoinEUIPrefixes               []types.EUI64Prefix                  `name:"join-eui-prefix" description:"JoinEUI prefixes handled by this JS"`
+	DefaultJoinEUI                types.EUI64                          `name:"default-join-eui" description:"the default JoinEUI for this JS"`
```
Comment (htdvisser) [inline pkg/joinserver/config.go]:
```suggestion
	DefaultJoinEUI                types.EUI64                          `name:"default-join-eui" description:"Default JoinEUI for this Join Server"`
```

Comment (htdvisser) [reply pkg/joinserver/config.go]: Please also change to "Join Server" above. We avoid shortening NS/AS/JS/... in places that users see.

**Review by htdvisser — APPROVED** [review]

File: cmd/internal/shared/joinserver/config.go [hunk 5]
```diff
@@ -24,4 +24,5 @@ var DefaultJoinServerConfig = joinserver.Config{
 	JoinEUIPrefixes: []types.EUI64Prefix{
 		{},
 	},
+	DefaultJoinEUI: types.EUI64{0x70, 0xb3, 0xd5, 0x7e, 0xd0, 0x00, 0x00, 0x00},
```
Comment (johanstokking) [inline cmd/internal/shared/joinserver/config.go]: This is not a good default value.

If we don't support not having a default value, it should be zero.

This value is reserved for The Things Join Server.

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 6]
```diff
@@ -455,6 +455,21 @@ var (
...
+							js, err := api.Dial(ctx, config.JoinServerGRPCAddress)
+							if err != nil {
+								return err
+							}
+							defaultJoinEUI, err := ttnpb.NewJsClient(js).GetDefaultJoinEUI(ctx, ttnpb.Empty)
+							if err != nil {
+								return err
+							}
+							logger.WithField("default_join_eui", defaultJoinEUI.JoinEui.String()).
+								Info("Successfully obtained Join Server's default Join EUI")
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]:
```suggestion
								Info("Successfully obtained Join Server's default JoinEUI")
```

File: cmd/internal/shared/joinserver/config.go (hunk 5 above)
Comment (KrishnaIyer) [reply cmd/internal/shared/joinserver/config.go]: I see. Ok that's fine with me. I've updated to just use zero.

**Review by johanstokking — APPROVED** [review]

**Pull request merged**